gcc/d/d-builtins.cc

```diff
@@ -69,21 +72,34 @@ static Type *
 build_dtype(tree type)
 {
   Type *dtype;
+  MOD mod = 0;
+
+  if (TYPE_READONLY (type))
+    mod |= MODconst;
+  if (TYPE_VOLATILE (type))
+    mod |= MODshared;
+
+  for (size_t i = 0; i < builtin_converted_syms.length(); ++i)
+    {
+      tree ti = builtin_converted_syms[i].ctype;
+      if (TYPE_MAIN_VARIANT (ti) == TYPE_MAIN_VARIANT (type))
+	return builtin_converted_syms[i].dtype;
+    }
 
   switch (TREE_CODE (type))
     {
     case POINTER_TYPE:
-      // Check for char * first. Needs to be done for chars/string.
-      if (TYPE_MAIN_VARIANT (TREE_TYPE (type)) == char_type_node)
-	return Type::tchar->pointerTo();
-
       dtype = build_dtype(TREE_TYPE (type));
       if (dtype)
 	{
+	  // Check for char * first. Needs to be done for chars/string.
+	  if (TYPE_MAIN_VARIANT (TREE_TYPE (type)) == char_type_node)
+	    return Type::tchar->addMod(dtype->mod)->pointerTo()->addMod(mod);
+
 	  if (dtype->ty == Tfunction)
-	    return new TypePointer(dtype);
-	  else
-	    return dtype->pointerTo();
+	    return (new TypePointer(dtype))->addMod(mod);
+
+	  return dtype->pointerTo()->addMod(mod);
```

Comment:
For reference (incase anyone looks at the PR from the distant future).

`addMod` applies the modifier in addition to the existing modifiers, whereas `castMod` removes shared and immutable attributes.